File: Cplusplus/topKFrequent.cpp
// Given a non-empty array of integers, return the k most frequent elements.

#include <iostream>
#include <vector>
#include <map>
#include <queue>

using namespace std;

class Compare {
public:
    bool operator() (pair<int, int> a, pair<int, int> b) {
        return a.second > b.second;
    }
};

vector<int> topKFrequent(vector<int>& nums, int k);

int main(int argc, char const *argv[])
{
    vector<int> nums = {1,1,1,2,2,3};
    int k = 2;
    vector<int> ans = topKFrequent(nums, k);
    for (int n : ans) {
        cout << n << endl;
    }
    return 0;
}

vector<int> topKFrequent(vector<int>& nums, int k)
{
    map<int, int> numFreq;

    for (int n : nums) {
        numFreq[n]++;
    }

    priority_queue<pair<int,int>, vector<pair<int, int>>, Compare> topK;

    auto it = numFreq.begin();

    while (it != numFreq.end()) {
        topK.push(make_pair (it->first, it->second));

        if (topK.size() > k) {
            topK.pop();
        }
        // cout << it->first << ": " << it->second << endl;
        it++;
    }

    vector<int> ans(k);

    int i = 0;

    while (!topK.empty()) {
        ans[i] = topK.top().first;
        topK.pop();
        i++;
    }

    return ans;
}

File: Golang/topKFrequent.go
// Given a non-empty array of integers, return the k most frequent elements.

package main

import (
	"container/heap"
)

// NumFreq store number and its frequency
type NumFreq struct {
	Num  int
	Freq int
}

// PriorityQueue array of NumFreq type
type PriorityQueue []*NumFreq

func (pq PriorityQueue) Len() int {
	return len(pq)
}

func (pq PriorityQueue) Less(i, j int) bool {
	return pq[i].Freq >= pq[j].Freq
}

func (pq PriorityQueue) Swap(i, j int) {
	pq[i], pq[j] = pq[j], pq[i]
}

// Push a value of type interface
func (pq *PriorityQueue) Push(x interface{}) {
	item := x.(*NumFreq)
	*pq = append(*pq, item)
}

// Pop a value from queue and return it
func (pq *PriorityQueue) Pop() interface{} {
	old := *pq
	n := len(*pq)
	item := old[n-1]
	// item.index = -1
	*pq = old[0 : n-1]
	return item
}

func topKFrequent(nums []int, k int) []int {
	numMap := make(map[int]int)

	for i := 0; i < len(nums); i++ {
		numMap[nums[i]]++
	}

	myNums := make(PriorityQueue, len(numMap))

	i := 0
	for num, freq := range numMap {
		myNums[i] = &NumFreq{num, freq}
		i++
	}

	heap.Init(&myNums)
	var ans []int
	for i = 0; i < k; i++ {
		num := heap.Pop(&myNums).(*NumFreq)
		ans = append(ans, num.Num)
	}

	return ans
}
